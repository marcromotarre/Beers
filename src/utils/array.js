export const generateAutoIncrementalArrayFromMinToMax = (min, max) => {
  return new Array(max - min + 1).fill().map((number, index) => min + index);
};
