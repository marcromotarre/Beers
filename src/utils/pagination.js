export const getNumPages = ({ totalElements, elementsPerPage }) => {
  return Math.ceil(totalElements / elementsPerPage);
};

export const generatePageNumbers = ({
  actualPage,
  numPages,
  maxNumShowedPages,
}) => {
  const pagesPerSide = Math.floor(maxNumShowedPages / 2);
  const arrayPages = new Array(numPages)
    .fill()
    .map((number, index) => index + 1);

  if (numPages <= maxNumShowedPages) return { pages: arrayPages };
  if (actualPage - pagesPerSide <= 1) {
    return {
      last: numPages,
      pages: arrayPages.filter((page) => page < maxNumShowedPages - 1),
    };
  }
  if (actualPage + pagesPerSide >= numPages) {
    return {
      first: 1,
      pages: arrayPages.filter(
        (page) => page > numPages - (maxNumShowedPages - 2)
      ),
    };
  }

  return {
    first: 1,
    pages: arrayPages.filter(
      (page) =>
        page < actualPage + (pagesPerSide - 1) &&
        page > actualPage - (pagesPerSide - 1)
    ),
    last: numPages,
  };
};
