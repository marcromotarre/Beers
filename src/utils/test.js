import React from 'react';
import { render } from '@testing-library/react';

import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from '@reach/router';

import { createStore } from 'redux';
import reducer from './../redux/reducers/index';
import { Provider } from 'react-redux';

export function renderRouteRedux(
  ui,
  {
    route = '/',
    history = createHistory(createMemorySource(route)),
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <LocationProvider history={history}>{children}</LocationProvider>
      </Provider>
    );
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export function renderRedux(
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export function renderReduxWithRouter(
  ui,
  {
    route = '/',
    history = createHistory(createMemorySource(route)),
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  return {
    ...render(
      <Provider store={store}>
        <LocationProvider history={history}>{ui}</LocationProvider>
      </Provider>
    ),
    history,
  };
}
