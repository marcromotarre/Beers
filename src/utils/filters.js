import { formatBrewed } from './text';
import { generateAutoIncrementalArrayFromMinToMax } from './array';

export const getBeersByBrewedFromTo = ({
  beers = [],
  brewedFrom,
  brewedTo,
}) => {
  return beers.filter((beer) => {
    const brewed = formatBrewed(beer.first_brewed);
    return (
      (brewed.year > brewedFrom.year && brewed.year < brewedTo.year) ||
      (brewed.year === brewedFrom.year &&
        brewed.month >= brewedFrom.month &&
        brewed.year < brewedTo.year) ||
      (brewed.year > brewedFrom.year &&
        brewed.year === brewedTo.year &&
        brewed.month <= brewedTo.month) ||
      (brewed.year === brewedFrom.year &&
        brewed.year === brewedTo.year &&
        brewed.month >= brewedFrom.month &&
        brewed.month <= brewedTo.month)
    );
  });
};

export const getMinBrewedDate = (beers = []) => {
  return beers
    .map((beer) => formatBrewed(beer.first_brewed))
    .sort(function (a, b) {
      return a.year <= b.year && a.month <= b.month ? -1 : 1;
    })[0];
};

export const getMaxBrewedDate = (beers) => {
  return beers
    .map((beer) => formatBrewed(beer.first_brewed))
    .sort(function (a, b) {
      return a.year <= b.year && a.month <= b.month ? 1 : -1;
    })[0];
};

export const getYearsFromMinToMaxBrewed = (beers) => {
  return generateAutoIncrementalArrayFromMinToMax(
    getMinBrewedDate(beers).year,
    getMaxBrewedDate(beers).year
  );
};

export const getMonths = () => {
  return generateAutoIncrementalArrayFromMinToMax(1, 12);
};
