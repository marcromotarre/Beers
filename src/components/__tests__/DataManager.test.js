import React from 'react';
import { renderRedux } from './../../utils/test';
import DataManager from '../DataManager';

/** this component should:
 *
 *      - manage content info.
 *          - show loader if redux beersReducer data is pending
 *          - show error if redux beersReducer data is error
 *          - show data if redux beersReducer data is success
 */

describe('DataManager Component', () => {
  it('DataManager Snapshot', () => {
    expect(
      renderRedux(
        <DataManager>
          <h1>data correct</h1>
        </DataManager>,
        {
          initialState: { beersReducer: { pending: false, error: false } },
        }
      ).baseElement
    ).toMatchSnapshot();

    expect(
      renderRedux(
        <DataManager>
          <h1>data correct</h1>
        </DataManager>,
        {
          initialState: { beersReducer: { pending: true, error: false } },
        }
      ).baseElement
    ).toMatchSnapshot();

    expect(
      renderRedux(
        <DataManager>
          <h1>data correct</h1>
        </DataManager>,
        {
          initialState: { beersReducer: { pending: false, error: true } },
        }
      ).baseElement
    ).toMatchSnapshot();
  });

  it('DataManager should show loader if there is pending', () => {
    const { container } = renderRedux(
      <DataManager>
        <h1>data correct</h1>
      </DataManager>,
      {
        initialState: { beersReducer: { pending: true, error: false } },
      }
    );
    const loadingImg = container.querySelector('img');
    expect(loadingImg.getAttribute('alt')).toBe('loading');
    expect(loadingImg.getAttribute('src')).toBe(
      'https://upload.wikimedia.org/wikipedia/commons/b/b1/Loading_icon.gif'
    );
  });

  it('DataManager should show children if is not pending and there is no error', () => {
    const { container } = renderRedux(
      <DataManager>
        <h1>data correct</h1>
      </DataManager>,
      {
        initialState: { beersReducer: { pending: false, error: false } },
      }
    );
    expect(container.querySelector('h1').innerHTML).toBe('data correct');
  });

  it('DataManager should show error message if there is error', () => {
    const { container } = renderRedux(
      <DataManager>
        <h1>data correct</h1>
      </DataManager>,
      {
        initialState: { beersReducer: { pending: false, error: true } },
      }
    );
    expect(container.querySelector('label').innerHTML).toBe(
      'Error Loading Data'
    );
  });
});
