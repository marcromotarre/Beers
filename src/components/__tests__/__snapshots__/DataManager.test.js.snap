// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataManager Component DataManager Snapshot 1`] = `
<body>
  <div>
    <h1>
      data correct
    </h1>
  </div>
</body>
`;

exports[`DataManager Component DataManager Snapshot 2`] = `
<body>
  <div>
    <h1>
      data correct
    </h1>
  </div>
  <div>
    <div
      class="sc-AxjAm kSkoow"
    >
      <img
        alt="loading"
        src="https://upload.wikimedia.org/wikipedia/commons/b/b1/Loading_icon.gif"
      />
    </div>
  </div>
</body>
`;

exports[`DataManager Component DataManager Snapshot 3`] = `
<body>
  <div>
    <h1>
      data correct
    </h1>
  </div>
  <div>
    <div
      class="sc-AxjAm kSkoow"
    >
      <img
        alt="loading"
        src="https://upload.wikimedia.org/wikipedia/commons/b/b1/Loading_icon.gif"
      />
    </div>
  </div>
  <div>
    <div
      class="sc-AxirZ klYqbc"
    >
      <label>
        Error Loading Data
      </label>
    </div>
  </div>
</body>
`;
